crypto/hpke/hpke.c

```diff
@@ -0,0 +1,3137 @@
+ * @param buf is the binary buffer with the (uncompressed) public value
+ * @param buflen is the length of the private key buffer
+ * @return a working EVP_PKEY * or NULL
+ */
+static EVP_PKEY * hpke_EVP_PKEY_new_raw_nist_public_key(OSSL_LIB_CTX *libctx,
+                                                        int curve,
+                                                        const char *gname,
+                                                        unsigned char *buf,
+                                                        size_t buflen)
+{
+    int erv = 1;
+    EVP_PKEY *ret = NULL;
+    EVP_PKEY_CTX *cctx = EVP_PKEY_CTX_new_from_name(libctx, "EC", NULL);
+
+    if (cctx == NULL) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    if (EVP_PKEY_paramgen_init(cctx) <= 0) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    if (EVP_PKEY_CTX_set_ec_paramgen_curve_nid(cctx, curve) <= 0) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    if (EVP_PKEY_paramgen(cctx, &ret) <= 0) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    if (EVP_PKEY_set1_encoded_public_key(ret, buf, buflen) != 1) {
+        EVP_PKEY_free(ret);
+        ret = NULL;
+        OSSL_HPKE_err;
+        goto err;
+    }
+
+err:
+    EVP_PKEY_CTX_free(cctx);
+    if (erv == 1)
+        return (ret);
+    else
+        return NULL;
+}
+
+/*
+ * There's an odd accidental coding style feature here:
+ * For all the externally visible functions in hpke.h, when
+ * passing in a buffer, the length parameter precedes the
+ * associated buffer pointer. It turns out that, entirely by
+ * accident, I did the exact opposite for all the static
+ * functions defined inside here. But since I was consistent
+ * in both cases, I'll declare that a feature and move on:-)
+ *
+ * For example, just below you'll see:
+ *          unsigned char *iv, size_t ivlen,
+ * ...whereas in hpke.h, you see:
+ *          size_t publen, unsigned char *pub,
+ */
+
+/*
+ * @brief do the AEAD decryption
+ *
+ * @param libctx is the context to use (normally NULL)
+ * @param suite is the ciphersuite
+ * @param key is the secret
+ * @param keylen is the length of the secret
+ * @param iv is the initialisation vector
+ * @param ivlen is the length of the iv
+ * @param aad is the additional authenticated data
+ * @param aadlen is the length of the aad
+ * @param cipher is obvious
+ * @param cipherlen is the ciphertext length
+ * @param plain is an output
+ * @param plainlen input/output, better be big enough on input, exact on output
+ * @return 1 for good otherwise bad
+ */
+static int hpke_aead_dec(OSSL_LIB_CTX *libctx,
+                         OSSL_HPKE_SUITE suite,
+                         unsigned char *key, size_t keylen,
+                         unsigned char *iv, size_t ivlen,
+                         unsigned char *aad, size_t aadlen,
+                         unsigned char *cipher, size_t cipherlen,
+                         unsigned char *plain, size_t *plainlen)
+{
+    int erv = 1;
```

Comment:
Here I would not use the `OSSL_HPKE_err` macro and just initialise `erv` to 0. Then just set it to 1 at the end of the function in the final success case.

Reply:
Fair enough, but seems a mere matter of style:-) The consistency of setting erv to 1 at the top of 'em all seems better to me TBH. (And the OSSL_HPKE_err macro's a gonner now too.)

Reply:
Agree with matt here.. This style is more error prone..